src/python/pants/core/goals/check.py

```diff
@@ -201,8 +202,29 @@ async def check(
         )
         for request_type in request_types
     )
+
+    request_to_field_set = [
+        (request, field_set) for request in requests for field_set in request.field_sets
+    ]
+
+    environment_names = await MultiGet(
+        Get(
+            EnvironmentName,
+            EnvironmentNameRequest,
+            EnvironmentNameRequest.from_field_set(field_set),
+        )
+        for (_, field_set) in request_to_field_set
+    )
+
+    request_to_env_name = {
+        (request, env_name)
+        for (request, _), env_name in zip(request_to_field_set, environment_names)
+    }
```

Comment:
If a request has multiple field sets, this might end up computing multiple environment names for one request, which will overwrite one another in this dict.

This code probably needs to either partition by environment name or error for that case? See #17328 on that topic: currently `mypy` internally partitions, but `check` itself does not.

Reply:
This is a literal for a set of `(request, env_name)` tuples, not a `dict`, so it'll run each request in each environment for which it is valid.

Reply:
I was considering whether this rule should actually re-construct each request with only the field sets that match each `EnvironmentName`, but for now, this is probably fine?

Reply:
> I was considering whether this rule should actually re-construct each request with only the field sets that match each `EnvironmentName`, but for now, this is probably fine?

Yea, that would effectively be partitioning, á la #17328. I expect that that will be necessary in the medium term.